diff --git a/shim/index.js b/shim/index.js
index 974854c2cd8ad8adc102db0c12a3880c41ee2529..0fa750f7ddda82935b5e9479e1c9b3ec9f0db1d8 100644
--- a/shim/index.js
+++ b/shim/index.js
@@ -1,7 +1 @@
-'use strict';
-
-if (process.env.NODE_ENV === 'production') {
-  module.exports = require('../cjs/use-sync-external-store-shim.production.js');
-} else {
-  module.exports = require('../cjs/use-sync-external-store-shim.development.js');
-}
+export { useSyncExternalStore } from 'react';
\ No newline at end of file
diff --git a/shim/with-selector.js b/shim/with-selector.js
index 91859dbd47282b19308a9b52e4afb897167ad0ec..ff1e30afadb90f0d1543c0f0b702d703b36674ea 100644
--- a/shim/with-selector.js
+++ b/shim/with-selector.js
@@ -1,7 +1,103 @@
-'use strict';
+/**
+ * @license React
+ * use-sync-external-store-with-selector.development.js
+ *
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
 
-if (process.env.NODE_ENV === 'production') {
-  module.exports = require('../cjs/use-sync-external-store-shim/with-selector.production.js');
-} else {
-  module.exports = require('../cjs/use-sync-external-store-shim/with-selector.development.js');
+import {
+  useSyncExternalStore,
+  useRef,
+  useEffect,
+  useMemo,
+  useDebugValue,
+} from 'react';
+
+function is(x, y) {
+  return (x === y && (0 !== x || 1 / x === 1 / y)) || (x !== x && y !== y);
+}
+
+const objectIs = typeof Object.is === 'function' ? Object.is : is;
+
+export function useSyncExternalStoreWithSelector(
+  subscribe,
+  getSnapshot,
+  getServerSnapshot,
+  selector,
+  isEqual
+) {
+  let instRef = useRef(null);
+  let inst;
+  if (null === instRef.current) {
+    inst = { hasValue: false, value: null };
+    instRef.current = inst;
+  } else {
+    inst = instRef.current;
+  }
+
+  instRef = useMemo(
+    function () {
+      function memoizedSelector(nextSnapshot) {
+        if (!hasMemo) {
+          hasMemo = true;
+          memoizedSnapshot = nextSnapshot;
+          const nextSelection = selector(nextSnapshot);
+          if (undefined !== isEqual && inst.hasValue) {
+            const currentSelection = inst.value;
+            if (isEqual(currentSelection, nextSelection)) {
+              return (memoizedSelection = currentSelection);
+            }
+          }
+          return (memoizedSelection = nextSelection);
+        }
+
+        const currentSelection = memoizedSelection;
+        if (objectIs(memoizedSnapshot, nextSnapshot)) {
+          return currentSelection;
+        }
+
+        const nextSelection = selector(nextSnapshot);
+        if (undefined !== isEqual && isEqual(currentSelection, nextSelection)) {
+          memoizedSnapshot = nextSnapshot;
+          return currentSelection;
+        }
+
+        memoizedSnapshot = nextSnapshot;
+        return (memoizedSelection = nextSelection);
+      }
+
+      let hasMemo = false;
+      let memoizedSnapshot;
+      let memoizedSelection;
+      const maybeGetServerSnapshot =
+        undefined === getServerSnapshot ? null : getServerSnapshot;
+      return [
+        function () {
+          return memoizedSelector(getSnapshot());
+        },
+        null === maybeGetServerSnapshot
+          ? undefined
+          : function () {
+              return memoizedSelector(maybeGetServerSnapshot());
+            },
+      ];
+    },
+    [getSnapshot, getServerSnapshot, selector, isEqual]
+  );
+
+  const value = useSyncExternalStore(subscribe, instRef[0], instRef[1]);
+
+  useEffect(
+    function () {
+      inst.hasValue = true;
+      inst.value = value;
+    },
+    [value]
+  );
+
+  useDebugValue(value);
+  return value;
 }
